psycopg3/transaction.py

```diff
@@ -0,0 +1,86 @@
+"""
+Transaction context managers returned by Connection.transaction()
+"""
+
+# Copyright (C) 2020 The Psycopg Team
+
+# TODO: Should this module be _underscored since it's implementation detail?
+from types import TracebackType
+from typing import Optional, Type, TYPE_CHECKING
+
+from .pq import TransactionStatus
+
+if TYPE_CHECKING:
+    from .connection import Connection
+
+
+class Transaction:
+    def __init__(self, conn: "Connection") -> None:
+        self._conn = conn
+        self._outer_transaction = False
```

Comment:
Maybe you can do without `_outer_transaction` because that's equivalent to having `_savepoint_name is None`?

Reply:
At the moment yes, but I was planning to implement the behaviour that if you specify savepoint name in outer Transaction() scope it will begin a transaction _and_ save a savepoint with your name, in which case there are two cases of "outer transaction" to differentiate.

This seems better than the other two options:
1. Ignore the savepoint name provided -- if a user is providing a name, we should respect it.
2. Raise exception saying "can't name outer transaction" -- this breaks composability; if I write a library that uses `with Transaction()` on a conn passed in, I shouldn't make restrictions on the state of that conn. Doing so brings us back to the past inconveniences of "you must pass me a connection with autocommit off for this code to work correctly" with the equivalent of "you must pass me a connection with an outer transaction already running for this code to be able to use `with Transaction(name=...)` correctly"!

So, TL;DR; true for now, but both will be needed soon :-)

Reply:
If you have a reason to leave it this way, that's cool for me: you know the path ahead better than myself on this feature :slightly_smiling_face: